podwatcher/extension.go

```diff
@@ -0,0 +1,112 @@
+package podwatcher
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"runtime"
+
+	eirinix "code.cloudfoundry.org/eirinix"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// https://github.com/cloudfoundry-incubator/eirini-staging/tree/master/image
+const (
+	defaultStagingExecutorEntrypoint   = "/packs/executor"
+	defaultStagingDownloaderEntrypoint = "/packs/downloader"
+	defaultStagingUploaderEntrypoint   = "/packs/uploader"
+	defaultRuntimeEntrypoint           = "/lifecycle/launch"
+	defaultFailGracePeriod             = "5"
+	defaultSuccessGracePeriod          = "5"
+)
+
+// GraceOptions lets customize the graceful periods and
+// the entrypoint of the images which are mutated to inject
+// the grace period logic
+type GraceOptions struct {
+	FailGracePeriod, SuccessGracePeriod string
+
+	StagingDownloaderEntrypoint string
+	StagingExecutorEntrypoint   string
+	StagingUploaderEntrypoint   string
+	RuntimeEntrypoint           string
+}
+
+// Extension changes pod definitions
+type Extension struct {
+	Logger  *zap.SugaredLogger
+	Options GraceOptions
+}
+
+// NewGracePeriodInjector returns the podwatcher extension which injects a grace Period on Eirini generated pods
+func NewGracePeriodInjector(opts *GraceOptions) *Extension {
+	if len(opts.StagingExecutorEntrypoint) == 0 {
+		opts.StagingExecutorEntrypoint = defaultStagingExecutorEntrypoint
+	}
+
+	if len(opts.StagingDownloaderEntrypoint) == 0 {
+		opts.StagingDownloaderEntrypoint = defaultStagingDownloaderEntrypoint
+	}
+
+	if len(opts.StagingUploaderEntrypoint) == 0 {
+		opts.StagingUploaderEntrypoint = defaultStagingUploaderEntrypoint
+	}
+
+	if len(opts.RuntimeEntrypoint) == 0 {
+		opts.RuntimeEntrypoint = defaultRuntimeEntrypoint
+	}
+
+	if len(opts.FailGracePeriod) == 0 {
+		opts.FailGracePeriod = defaultFailGracePeriod
+	}
+
+	if len(opts.SuccessGracePeriod) == 0 {
+		opts.SuccessGracePeriod = defaultSuccessGracePeriod
+	}
+
+	return &Extension{Options: *opts}
+}
+
+// Handle injects gracefulPeriod in opi containers:
+// In case an application is too fast to fail, we might miss logs as Kubernetes will terminate the container
+// before the watcher had occasion to attach to stream any logs.
+// - Finalizers let the pod terminates, not giving any chance to get logs from terminated container.
+// - PostStop/PreStart hooks doesn't guarantee execution in case of containers terminated by failures (e.g. an invalid app was pushed in Eirini), if causing crashloopbackoff
+//   hooks aren't executed correctly.
+func (ext *Extension) Handle(ctx context.Context, eiriniManager eirinix.Manager, pod *corev1.Pod, req admission.Request) admission.Response {
```

Comment:
Would it make sense to filter on labels / annotations to skip attempting to process pods that are not from eirini? Mostly as a check against unintentionally modifying unrelated pods, as opposed to speed optimization.

Reply:
Yes it would, but this is a feature request that come up recently while creating other extensions, this probably is a feature that should be bundled in EiriniX to avoid boilerplate code in all the extensions depending on it.